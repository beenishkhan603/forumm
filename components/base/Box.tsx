import { Theme, useTheme } from '@libs/useTheme'
import {
  CSSProperties,
  DetailedHTMLProps,
  HTMLAttributes,
  LegacyRef,
} from 'react'
import LoadingSpinner from './LoadingSpinner'

const Box = (
  props: {
    children?: any
    loading?: boolean
    color?: keyof Theme
    show?: boolean
    ignoreTheme?: boolean
    textColour?: keyof Theme
    blur?: boolean
    innerRef?: LegacyRef<HTMLDivElement>
  } & DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>
) => {
  const { theme } = useTheme()
  const {
    children,
    loading = false,
    color,
    textColour,
    show = true,
    ignoreTheme = false,
    style,
    blur = false,
    innerRef,
    ...divProps
  } = props
  if (!show) {
    return null
  }
  return (
    <div
      {...divProps}
      style={
        ignoreTheme || style
          ? style ?? {}
          : {
            backgroundColor: color && theme ? theme[color] : undefined,
            color:
              theme && textColour
                ? theme[textColour]
                : theme.textColour ?? undefined,
          }
      }
      ref={innerRef}
    >
      {loading ? (
        <Box
          data-test-id={'box-spinner'}
          className="w-full h-full justify-center items-center flex"
        >
          <LoadingSpinner size="medium" />
        </Box>
      ) : blur ? (
        <span style={{ filter: 'blur(2px)' }}>{children}</span>
      ) : (
        children
      )}
    </div>
  )
}

export default Box
