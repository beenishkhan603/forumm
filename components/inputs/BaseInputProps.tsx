export interface BaseInputProps<T = any> {
  /**
   * The label for the input.
   */
  label?: string
  /**
   * The placeholder text for the input.
   */
  placeholder?: string
  /**
   * Class name to apply to the outer wrapper.
   */
  className?: string
  /**
   * The value of the input.
   */
  value?: T
  /**
   * Callback with value when the input value changes.
   */
  onChange?: (value: T, isInvalid?: boolean | undefined) => void
  /**
   * Input required for form submission
   */
  required?: boolean
  /**
   * Show hint next to input label
   */
  hint?: string
  /**
   * Validation rules for the input
   */
  validations?: Validations
  filterDate?: (date: Date) => boolean
  filterTime?: (date: Date) => boolean
  disabled?: boolean
  testid?: string
  show?: boolean
  labelBgColour?: string
  textColour?: string
  border?: string
  isProtected?: boolean
  tooltip?: string
}

export type Validations = {
  minLength?: {
    value: number
    message: string
  }
  maxLength?: {
    value: number
    message: string
  }
  pattern?: {
    value: RegExp
    message: string
  }
  isURL?: boolean
}
