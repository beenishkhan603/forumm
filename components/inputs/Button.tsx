import Box from '@components/base/Box'
import LoadingSpinner from '@components/base/LoadingSpinner'
import Modal from '@components/base/Modal'
import { getParentBg, isReadable, Theme, useTheme } from '@libs/useTheme'
import { getContrastColor } from '@libs/Utility/util'
import { motion, Variants } from 'framer-motion'
import React, { RefObject, useEffect, useRef, useState } from 'react'
import { IconContext, IconType } from 'react-icons'

interface ButtonProps {
  /**
   * Is this the principal call to action on the page?
   */
  type?:
    | 'primary'
    | 'secondary'
    | 'tertiary'
    | 'special'
    | 'danger'
    | 'success'
    | 'link'
    | 'modal'
    | 'donation'
    | 'create'
    | 'square'
    | 'blackbaud'
  /**
   * What background color to use
   */
  backgroundColor?: string
  /**
   * What text color to use
   */
  textColor?: string
  /**
   * How large should the button be?
   */
  size?: 'small' | 'medium' | 'large' | 'auto' | 'full'
  /**
   * Button contents
   */
  title?: string
  /**
   * icon
   */
  icon?: JSX.Element
  iconPos?: 'start' | 'end'
  /**
   * Optional click handler
   */
  onClick?: (e: React.MouseEvent<HTMLElement>) => void
  onMouseDown?: (e: React.MouseEvent<HTMLElement>) => void
  /**
   * Optional class name
   */
  className?: string
  /**
   * Optional href link
   */
  href?: string
  confirmationModal?: { title: string; content: string }
  buttonType?: 'button' | 'submit' | 'reset'
  loading?: boolean
  disabled?: boolean
  show?: boolean
  textColour?: keyof Theme | string
  style?: any
  isSelected?: boolean
  isInConfirmationModal?: boolean
  iconColor?: string
}

const buttonVariants: Variants = {
  hover: {
    scale: 1,
  },
}

/**
 * Primary UI component for user interaction
 */
function Button({
  type = 'primary',
  size = 'medium',
  backgroundColor,
  textColor,
  title,
  icon,
  iconPos = 'start',
  href,
  className,
  buttonType = 'button',
  loading,
  disabled,
  confirmationModal,
  onClick,
  onMouseDown,
  show = true,
  textColour,
  isSelected = false,
  isInConfirmationModal = false,
  iconColor,
  ...props
}: ButtonProps) {
  const { theme } = useTheme()
  const isLight = theme.type === 'LIGHT'

  const sizeClasses = {
    small: 'px-4 py-4 text-xs h-6',
    medium: 'px-6 py-4 text-base h-8',
    large: 'px-8 py-4 text-base h-10',
    full: 'px-8 py-4 text-base h-10 w-full',
  }
  const colorClasses = {
    primary: `${
      isLight
        ? 'bg-forumm-blue-2 text-white relative'
        : 'bg-forumm-blue text-white hover:bg-blue-700 relative'
    }`,
    secondary: `${
      isLight
        ? 'border border-forumm-menu-border text-forumm-blue-2 bg-forumm-white relative'
        : 'border-2 border-panel-gray text-midnight-light2 relative'
    } `,
    tertiary: 'bg-tertiary border',
    square: 'bg-tertiary border',
    special: 'border-forumm-orange border text-forumm-orange',
    blackbaud: 'border-blackbaud border text-blackbaud',
    danger: 'border-red-600 text-red-600 border',
    success: 'border-green-500 text-green-500 border',
    disabled: `${
      isLight
        ? '!bg-disabled-bg text-disabled-text border-disabled'
        : 'border-2 border-panel-gray text-forumm-gray'
    }`,
    link: `${
      isLight ? 'text-forumm-blue-2 bg-white' : 'text-midnight-light2'
    } `,
    modal: 'border border-panel-gray',
    donation: 'border border-tertiary',
    create: 'text-forumm-blue',
  }

  const buttonRef = useRef<HTMLButtonElement | HTMLAnchorElement>(null)
  const [computedStyle, setComputedStyle] = useState<
    CSSStyleDeclaration & { parentBg?: string }
  >()

  const [showConfirmation, setShowConfirmation] = useState(false)

  const handleAutoSize = (
    content: string | any[] | JSX.Element | null | undefined
  ) => {
    if (
      typeof content === 'string' &&
      content.length > 15 &&
      window.innerWidth < 1012
    )
      return sizeClasses['small']
    if (window.innerWidth < 640) return sizeClasses['small']
    if (window.innerWidth <= 960) return sizeClasses['medium']
    if (window.innerWidth > 960) return sizeClasses['large']
  }

  useEffect(() => {
    if (buttonRef.current) {
      const computed = window.getComputedStyle(buttonRef.current)
      const parentBg = getParentBg(buttonRef.current)
      setComputedStyle({ ...computed, parentBg })
    }
  }, [buttonRef.current])

  const computedBg = computedStyle?.backgroundColor ?? ''
  const computedText = computedStyle?.color ?? ''

  const textColorForTransparentBg = isReadable(
    computedText,
    computedStyle?.parentBg ?? ''
  )
    ? computedText
    : getContrastColor(computedStyle?.parentBg ?? '#ffffff')

  const buttonBgColor = backgroundColor
    ? backgroundColor
    : type === 'primary'
    ? theme.highlightColour ?? ''
    : computedBg

  const bgIsTransparent = [buttonBgColor, computedBg].includes(
    'rgba(0, 0, 0, 0)' || type === 'tertiary'
  )
  const contrastColor = getContrastColor(buttonBgColor)

  const tertiaryTextColor = bgIsTransparent
    ? textColorForTransparentBg
    : contrastColor

  let buttonTextColor = textColor
    ? isReadable(textColor, buttonBgColor)
      ? textColor
      : tertiaryTextColor
    : isReadable(computedText, buttonBgColor)
    ? computedText
    : tertiaryTextColor

  if (
    !buttonBgColor &&
    (/^((?!chrome|android).)*safari/i.test(navigator.userAgent) ||
      /firefox/i.test(navigator.userAgent))
  ) {
    buttonTextColor = '#000'
  }

  // if (title === 'Confirm')
  //   console.log({
  //     input: {
  //       title,
  //       icon,
  //       textColor,
  //       backgroundColor,
  //     },
  //     output: {
  //       computedBg,
  //       computedText,
  //       textColorForTransparentBg,
  //       buttonBgColor,
  //       buttonTextColor,
  //       contrastColor,
  //       tertiaryTextColor,
  //       bgIsTransparent,
  //     },
  //   })

  const getIcon = () => {
    if (!icon) return null

    // TODO: implement computed contrast for icon colour.
    return (
      <IconContext.Provider
        value={{
          style: {
            fill: iconColor ? iconColor : theme.textColour,
          },
        }}
      >
        {icon}
      </IconContext.Provider>
    )
  }

  const getTextContent = () => {
    if (loading) return <LoadingSpinner />
    if (title && icon)
      return (
        <span className="flex items-center space-between">
          {iconPos === 'start' && getIcon()}
          <p className="m-0 px-1 text-sm">{title}</p>
          {iconPos === 'end' && getIcon()}
        </span>
      )
    return title ?? getIcon()
  }

  if (href) {
    return (
      <motion.a
        ref={buttonRef as RefObject<HTMLAnchorElement>}
        variants={buttonVariants}
        whileHover="hover"
        whileTap="hover"
        className={`
        ${colorClasses[type]}         
        btn hover:animate-heartbeat rounded-full flex justify-center items-center text-center ${
          size !== 'auto' ? sizeClasses[size] : handleAutoSize(title)
        } my-0 ${className}`}
        style={{
          backgroundColor,
        }}
        href={href}
      >
        {title ?? getIcon()}
      </motion.a>
    )
  }

  return (
    <>
      {confirmationModal && (
        <Modal
          title={confirmationModal?.title}
          show={showConfirmation}
          setShow={setShowConfirmation}
        >
          <Box>{confirmationModal?.content}</Box>
          <Box className="flex w-full space-x-4 mt-4">
            <Button
              onClick={() => setShowConfirmation(false)}
              className="flex-1"
              title="Cancel"
              type="tertiary"
              style={{
                backgroundColor: 'transparent',
              }}
            />
            <Button
              onClick={(e) => {
                e.stopPropagation()
                if (onClick) onClick(e)
                setShowConfirmation(false)
              }}
              className="flex-1"
              title="Confirm"
              type="danger"
              style={{
                backgroundColor: 'transparent',
              }}
            />
          </Box>
        </Modal>
      )}
      {show && (
        <motion.button
          ref={buttonRef as RefObject<HTMLButtonElement>}
          variants={disabled ? undefined : buttonVariants}
          whileHover="hover"
          whileTap="hover"
          type={buttonType}
          className={`
      ${colorClasses[disabled ? 'disabled' : type]} ${
            type === 'square' ? 'rounded' : 'rounded-full'
          }  ${
            size !== 'auto'
              ? sizeClasses[size]
              : handleAutoSize(getTextContent())
          } flex items-center hover:animate-heartbeat justify-center ${className}`}
          style={
            buttonBgColor
              ? {
                  backgroundColor: buttonBgColor,
                }
              : {}
          }
          onMouseDown={onMouseDown}
          onClick={(e) => {
            e.stopPropagation()
            if (disabled) return
            if (confirmationModal) {
              setShowConfirmation(true)
              return
            }
            return onClick ? onClick(e) : null
          }}
          disabled={disabled}
          {...props}
        >
          <span
            className={`relative z-10`}
            style={
              buttonTextColor
                ? {
                    color: buttonTextColor,
                  }
                : {}
            }
          >
            {getTextContent()}
          </span>
        </motion.button>
      )}
    </>
  )
}

export { Button }
