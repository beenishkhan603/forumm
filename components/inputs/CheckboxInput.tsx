import Box from '@components/base/Box'
import Text from '@components/base/Text'
import Tooltip from '@components/tootilp/Tooltip'
import { useTheme } from '@libs/useTheme'
import React from 'react'
import { BaseInputProps } from './BaseInputProps'
import { validate } from './Validations'

export interface CheckboxInputProps extends BaseInputProps {
  children: JSX.Element
  border?: string
}

/**
 * Primary UI component for text input
 */
export const CheckboxInput = ({
  label,
  show = true,
  placeholder,
  value = '',
  onChange,
  className,
  hint,
  validations,
  required,
  testid,
  children,
  border = 'border',
  isProtected = true,
}: CheckboxInputProps) => {
  const { theme } = useTheme()
  const errorMessage = validate(value, validations)

  const backgroundColour = theme.editorBackgroundColour

  if (!show) return <></>
  return (
    <fieldset
      className={`rounded-2xl transition-all bg-transparent ${border} ${
        errorMessage ? 'text-red-500 border-red-500' : ''
      } rounded text-sm outline-none placeholder-gray-500 ${className}`}

      // className={[
      //   'flex flex-col space-y-1 border border-black',
      //   className,
      // ].join(' ')}
    >
      {label && border?.length > 0 && (
        <legend
          className={`text-sm text-justify ml-2 z-10 ${backgroundColour} ${
            required ? 'pl-2' : 'px-2'
          } transition-all ${errorMessage ? 'text-red-500' : ''}`}
        >
          <Tooltip show={required} tooltip="Required to publish.">
            <span
              className={`z-10 ${
                required
                  ? "after:content-['*'] after:mx-2 after:text-red-500"
                  : ''
              }`}
            >
              {label}
            </span>
          </Tooltip>
        </legend>
      )}
      <Box className="flex items-center justify-center w-full p-2 px-4">
        <input
          data-testid={testid ? `${testid}` : null}
          type="checkbox"
          className="h-4 w-4 text-forumm-blue border-gray-300 inline-block"
          name={label}
          checked={value === true}
          onClick={(e) => {
            e.stopPropagation()
          }}
          onChange={(e) => {
            // e.stopPropagation()
            // e.preventDefault()
            onChange && onChange(e.target.checked, false)
          }}
        />
        <Box className="inline-block ml-2">{children}</Box>
      </Box>

      <Box
        ignoreTheme
        className={`transition-all ${
          errorMessage ? 'opacity-100' : 'opacity-0'
        } text-red-500 text-sm font-bold`}
      >
        {errorMessage ?? '-'}
      </Box>
    </fieldset>
  )
}
