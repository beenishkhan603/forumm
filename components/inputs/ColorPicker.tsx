import Box from '@components/base/Box'
import { tryGetReadableColour, useTheme } from '@libs/useTheme'
import ColorPickerIcon from '@public/images/ColorPicker.svg'
import { motion } from 'framer-motion'
import React, { useEffect, useRef, useState } from 'react'
import { SwatchesPicker } from 'react-color'
import { BaseInputProps } from './BaseInputProps'
import { validate } from './Validations'

export interface TextInputProps extends BaseInputProps {
  /**
   * The type of the input.
   */
  type?: 'text' | 'password' | 'email' | 'tel' | 'number'
}

/**
 * Primary UI component for text input
 */
export const ColorPicker = ({
  label,
  placeholder,
  value = '',
  onChange,
  className,
  hint,
  validations,
  required,
  type = 'text',
  testid,
}: TextInputProps) => {
  const { theme } = useTheme()
  const colourPickerButtonRef = useRef<HTMLImageElement>(null)
  const errorMessage = validate(value, validations)
  const [showPickColor, setShowPickColor] = useState(false)
  const [pickColor, setPickColor] = useState(value)
  const [contrastColor, setContrastColor] = useState(theme.textColour)

  const parentHasClassRecursive = (
    element: any,
    classname: string
  ): boolean => {
    try {
      if (element?.className?.split(' ').indexOf(classname) >= 0) return true
    } catch (e) {}
    return (
      (element.parentNode &&
        parentHasClassRecursive(element.parentNode, classname)) ??
      false
    )
  }

  useEffect(() => {
    setPickColor(value ?? theme.backgroundColour)
    setContrastColor(tryGetReadableColour(value ?? theme.backgroundColour))
  }, [value, theme.backgroundColour])

  useEffect(() => {
    function handleClick(event: MouseEvent) {
      const clickedOnColourPicker = parentHasClassRecursive(
        event.target,
        'swatches-picker'
      )
      if (
        colourPickerButtonRef.current?.contains(event.target as Node) ||
        clickedOnColourPicker
      ) {
        return
      }
      setShowPickColor(false)
    }
    document.addEventListener('mousedown', handleClick)
    return () => {
      document.removeEventListener('mousedown', handleClick)
    }
  })

  const availableColors = [
    // Colors for good contrast with white text
    ['#8B0000', '#A52A2A', '#cc0a0a', '#DC143C'],
    ['#ffbb00', '#ed9315', '#cf7202', '#a35802'],
    ['#9ACD32', '#6B8E23', '#556B2F', '#808000'],
    ['#008000', '#228B22', '#078507', '#2E8B57'],
    ['#008B8B', '#20B2AA', '#5F9EA0', '#008080'],
    ['#00008B', '#0000CD', '#000080', '#12124f'],
    ['#4B0082', '#800080', '#8A2BE2', '#9932CC'],
    ['#C71585', '#DB7093', '#FF1493', '#ad075a'],
    ['#A52A2A', '#8B4513', '#ff793b', '#D2691E'],
    ['#696969', '#808080', '#A9A9A9', '#778899'],

    // Colors for good contrast with black text
    ['#fc4732', '#a86d59', '#b32525', '#6e0b0b'],
    ['#ffbb00', '#d4945b', '#fa8150', '#fa4e0f'],
    ['#ffff05', '#d4be00', '#8a8a11', '#9c8700'],
    ['#00de95', '#21b853', '#4b734b', '#227a5d'],
    ['#00f5f5', '#14b8b8', '#0a9191', '#207173'],
    ['#407c8f', '#2890d1', '#0a4963', '#145259'],
    ['#6363ff', '#5c2ef2', '#5c13c2', '#4f0599'],
    ['#ef42ff', '#9e649e', '#cc7ecc', '#b015aa'],
    ['#73215a', '#7d1047', '#b80d40', '#573e3e'],
    ['#6a7f9c', '#8f8f8f', '#707070', '#97aac2'],
  ]

  const input = (
    <Box className="relative">
      {showPickColor ? (
        <SwatchesPicker
          width={250}
          className="absolute top-full right-0 z-50"
          color={pickColor}
          onChange={(color) => {
            setPickColor(color.hex)
            onChange && onChange(color.hex, false)
          }}
          colors={availableColors}
        />
      ) : (
        ''
      )}
      {pickColor ? (
        <Box
          className="w-5 h-5 cursor-pointer absolute my-4 mr-4 right-0 rounded-full border-white border-solid border-2 drop-shadow shadow-inner"
          innerRef={colourPickerButtonRef}
          style={{ backgroundColor: pickColor }}
          onClick={() => setShowPickColor(!showPickColor)}
        ></Box>
      ) : (
        <ColorPickerIcon
          className="w-5 h-5 cursor-pointer absolute my-4 mr-4 right-0"
          onClick={() => setShowPickColor(!showPickColor)}
        />
      )}

      <input
        required={required}
        value={value}
        onChange={(e) => {
          const isInvalid = validate(e.target.value, validations) !== ''
          onChange && onChange(e.target.value, isInvalid)
        }}
        onFocus={() => setShowPickColor(true)}
        type={type}
        autoComplete="on"
        placeholder={value ? '' : 'Select Color'}
        data-testid={testid ?? null}
        style={{
          backgroundColor: pickColor,
          color: contrastColor,
        }}
        className={`bg-transparent select-none cursor-pointer ${
          errorMessage && 'text-red-500 border-red-500'
        } rounded px-4 py-4 text-sm outline-none w-full`}
      />
    </Box>
  )

  return (
    <label className={['flex flex-col space-y-1', className].join(' ')}>
      <span className={`transition-all ${errorMessage ? 'text-red-500' : ''}`}>
        {label} <span className="ml-2 text-xs">{hint}</span>
      </span>
      {input}
      <Box
        className={`transition-all ${
          errorMessage ? 'opacity-100' : 'opacity-0'
        } text-red-500 text-sm font-bold`}
      >
        {errorMessage}
      </Box>
    </label>
  )
}
