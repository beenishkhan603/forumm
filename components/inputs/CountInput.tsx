import Box from '@components/base/Box'
import React, { useEffect, useState } from 'react'
import { BaseInputProps } from './BaseInputProps'
import { validate } from './Validations'
import { useTheme } from '@libs/useTheme'

export interface PriceInputProps extends BaseInputProps {
  currencySymbol?: string | '£' | '$' | '€' | '¥' | '₹'
  minValue?: number
  maxValue?: number
}

/**
 * Primary UI component for text input
 */
export const CountInput = ({
  label,
  value = 0,
  onChange,
  className,
  validations,
  required,
  labelBgColour,
  textColour,
  border = 'border',
  show = true,
  minValue = 0,
  maxValue = 99,
}: PriceInputProps) => {
  const { theme } = useTheme()
  const errorMessage = validate(value, validations)
  const [count, setCount] = useState<number>(value)

  useEffect(() => {
    if (onChange) onChange(count)
  }, [count])

  if (!show) return <></>

  const input = (
    <input
      required={required}
      value={count}
      onChange={(e) => {
        let val = parseInt(e.target.value, 10)
        onChange && onChange(val)
        setCount(val > maxValue ? maxValue : val < minValue ? minValue : val)
      }}
      type="number"
      className={`transition-all bg-transparent border-t border-b cursor-pointer ${
        errorMessage ? 'text-red-500 border-red-500' : ''
      } py-4 text-sm placeholder-gray-500 px-1 w-10`}
    />
  )

  const bgHint = theme.type === 'DARK' ? 'bg-midnight-dark' : 'bg-white'
  const backgroundColour =
    labelBgColour ?? theme?.backgroundColourSecondary ?? '#0D0E11'

  return (
    <label
      className={['flex flex-row justify-between items-center', className].join(
        ' '
      )}
      style={{ color: textColour }}
    >
      {label && (
        <span
          className={`text-sm z-10 ${backgroundColour} ${
            required ? 'pl-2' : 'px-2'
          } transition-all ${errorMessage ? 'text-red-500' : ''} flex-3`}
        >
          <span
            style={{ backgroundColor: backgroundColour }}
            className={`z-10 ${
              required
                ? "after:content-['*'] after:mx-2 after:text-red-500"
                : ''
            }`}
          >
            {label}
          </span>
        </span>
      )}
      <Box className="flex flex-row flex-1 justify-end">
        <Box
          className="font-bold w-6 h-full border rounded-l-lg py-[14px] flex justify-center cursor-pointer "
          onClick={() => {
            setCount((state) => (state > minValue ? state - 1 : state))
          }}
        >
          -
        </Box>
        {input}
        <Box
          className="font-bold w-6 h-full border rounded-r-lg py-[14px] flex justify-center cursor-pointer"
          onClick={() => {
            setCount((state) => (state < maxValue ? state + 1 : state))
          }}
        >
          +
        </Box>
      </Box>
      <Box
        ignoreTheme
        className={`transition-all ${
          errorMessage ? 'opacity-100' : 'opacity-0'
        } text-red-500 text-sm font-bold`}
      >
        {errorMessage ?? '-'}
      </Box>
    </label>
  )
}
