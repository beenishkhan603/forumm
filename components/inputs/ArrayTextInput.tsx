import { Autocomplete } from '@aws-amplify/ui-react'
import Box from '@components/base/Box'
import React, { useEffect, useState } from 'react'
import { AiFillCloseCircle } from 'react-icons/ai'
import { useTheme } from '@libs/useTheme'
import { BaseInputProps } from './BaseInputProps'
import { Button } from './Button'
import { validate } from './Validations'
import Tooltip from '@components/tootilp/Tooltip'

export interface ArrayTextInputProps
  extends BaseInputProps<{ key: string; value: string }[]> {
  /**
   * The type of the input.
   */
  onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void
  border?: 'border' | 'border-none'
  max?: number
  min?: number
  options: { id: string; label: string }[]
}

/**
 * Primary UI component for text input
 */
export const ArrayTextInput = ({
  label,
  placeholder,
  value = [],
  onChange,
  className,
  hint,
  validations,
  required,
  border = 'border',
  onKeyDown,
  max,
  min,
  options,
  labelBgColour,
  tooltip,
}: ArrayTextInputProps) => {
  const { theme } = useTheme()
  const errorMessage = validate(value, validations)

  const [showTooltip, setShowTooltip] = useState(true)
  useEffect(() => {
    // Autohide tooltip if the input is not required.
    if (tooltip?.toLowerCase().includes('require') && !required)
      setShowTooltip(false)
  }, [tooltip, required])

  const [textInput, setTextInput] = useState('')
  const [selectedVal, setSelectedVal] = useState<{
    key: string
    value: string
  }>()

  const input = (
    <Autocomplete
      label={label}
      value={textInput}
      required={required}
      onKeyDown={onKeyDown}
      onChange={(e) => {
        setTextInput(e.target.value)
      }}
      onSelect={(e) => {
        setSelectedVal({ key: e.label, value: e.id })
        setTextInput(e.label)
      }}
      innerEndComponent={null}
      outerEndComponent={
        <Button
          onClick={() => {
            if (
              selectedVal &&
              value.findIndex((v) => v.value === selectedVal.value) === -1
            ) {
              onChange?.([...value, selectedVal])
              setTextInput('')
            }
          }}
          title="Add"
          className="h-[54px] ml-2"
          type="tertiary"
        />
      }
      max={max}
      min={min}
      placeholder={placeholder}
      hasSearchIcon={false}
      inputStyles={{
        padding: '15.5px 16px',
        fontSize: '14px',
        color: theme.textColour,
      }}
      options={options}
      crossOrigin={undefined}
      onPointerEnterCapture={undefined}
      onPointerOutCapture={undefined}
      onPointerLeaveCapture={undefined}
    />
  )

  return (
    <>
      <label
        style={{ color: 'initial' }}
        className={['relative flex-col space-y-1', className].join(' ')}
      >
        {label && (
          <Tooltip
            show={showTooltip}
            tooltip={tooltip}
            className={`flex flex-col`}
          >
            <span
              className={`transition-all text-sm pl-1 pr-1 absolute -top-3 left-2 z-10 ${
                errorMessage ? 'text-red-500' : ''
              }`}
              style={{
                backgroundColor: labelBgColour ?? theme.backgroundColour,
                color: theme.textColour,
              }}
            >
              {label}
              <span className="ml-2 text-xs">{hint}</span>
            </span>
          </Tooltip>
        )}
        {input}
        <Box
          ignoreTheme
          className={`transition-all ${
            errorMessage ? 'opacity-100' : 'opacity-0'
          } text-red-500 text-sm font-bold`}
        >
          {errorMessage}
        </Box>
      </label>
      <div className="flex flex-wrap">
        {value.map((v, i) => (
          <div
            key={i}
            className="bg-white text-black rounded px-2 m-1 flex items-center space-x-1"
          >
            <span>{v.key}</span>
            <AiFillCloseCircle
              className="hover:opacity-50 cursor-pointer"
              onClick={() => onChange?.(value.filter((i) => i !== v))}
            />
          </div>
        ))}
      </div>
    </>
  )
}
