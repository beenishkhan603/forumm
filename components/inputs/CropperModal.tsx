import { uploadFile } from '@aws-amplify/ui'
import Modal from '@components/base/Modal'
import { getCroppedImg } from '@libs/imageUtils'
import { useAnimationFrame } from 'framer-motion'
import { useCallback, useEffect, useState } from 'react'
import Cropper, { Area, Point } from 'react-easy-crop'
import { Button } from './Button'
import { v4 } from 'uuid'
import Box from '@components/base/Box'

export interface CropperModalProps {
  originalImageSrc?: string
  setOriginalImageSrc: (src?: string, file?: File) => Promise<void>
  show: boolean
  setShow: (show: boolean) => void
  aspectRatio?: number
  cropShape?: 'rect' | 'round'
}

export const CropperModal = ({
  originalImageSrc,
  setOriginalImageSrc,
  show,
  setShow,
  aspectRatio = 1,
  cropShape = 'rect',
}: CropperModalProps) => {
  const [crop, setCrop] = useState<Point>({ x: 0, y: 0 })
  const [zoom, setZoom] = useState(1)
  const [croppedImage, setCroppedImage] = useState<{
    url: string
    blob: Blob | null
  } | null>()
  const [animating, setAnimating] = useState(true)

  const onCropComplete = useCallback(
    async (croppedArea: Area, croppedAreaPixels: Area) => {
      const croppedImage = await getCroppedImg(
        originalImageSrc!,
        croppedAreaPixels,
        0
      )
      setCroppedImage(croppedImage)
    },
    [originalImageSrc]
  )
  return (
    <Modal
      show={originalImageSrc != null && show}
      setShow={async () => {
        setShow(false)
        await setOriginalImageSrc()
      }}
      animating={(animating) => {
        setAnimating(animating)
      }}
    >
      <Box className="relative h-[60vh] w-full">
        {!animating && (
          // Started causing a build error out of the blue, works fine.
          // @ts-ignore
          <Cropper
            image={originalImageSrc}
            crop={crop}
            zoom={zoom}
            aspect={aspectRatio}
            onCropChange={setCrop}
            onCropComplete={onCropComplete}
            onZoomChange={setZoom}
            classes={{ containerClassName: 'z-50' }}
            cropShape={cropShape}
          />
        )}
      </Box>
      <Button
        title="Crop"
        className="w-full mt-4"
        onClick={async () => {
          if (croppedImage && croppedImage.blob) {
            const file = new File([croppedImage.blob], `${v4()}.jpg`, {
              type: 'image/jpeg',
            })
            await setOriginalImageSrc(croppedImage.url, file)
          } else {
            await setOriginalImageSrc(originalImageSrc)
          }
          setShow(false)
        }}
      />
    </Modal>
  )
}
