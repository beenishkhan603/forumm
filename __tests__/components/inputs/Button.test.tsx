import Box from '@components/base/Box'
import { Button } from '@components/inputs/Button'
import { render } from '@testing-library/react'

const BG_FORUM_BLUE = 'background-color: rgb(55, 99, 233)'

describe('Button Test Suite', () => {
  it('Button test', () => {
    const { container } = render(
      <Box className="bg-black">
        <Button
          show={true}
          type="primary"
          className={''}
          iconColor={''}
          backgroundColor={''}
          textColor={''}
        />
      </Box>
    )
    const button = container.getElementsByTagName('button')[0]

    console.log(button)

    expect(button).toBeInTheDocument()
    expect(button).toHaveStyle(BG_FORUM_BLUE)
    // expect(button).toHaveStyle('color: ButtonText;')

    expect(container).toMatchSnapshot()
  })

  it('Button test', () => {
    const { container } = render(
      <Box className="bg-black">
        <Button
          show={true}
          className={undefined}
          iconColor={undefined}
          backgroundColor={'#f1f1f1'}
          textColor={'#1F1F1F'}
        />
      </Box>
    )
    const button = container.getElementsByTagName('button')[0]

    console.log({ button, style: button.style })

    expect(button).toBeInTheDocument()
    expect(button).not.toHaveStyle(BG_FORUM_BLUE)
    // expect(button).toHaveStyle('color: #1F1F1F')

    expect(container).toMatchSnapshot()
  })
})
