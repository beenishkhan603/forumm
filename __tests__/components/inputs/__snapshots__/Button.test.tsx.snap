// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Test Suite Button test 1`] = `
<div>
  <div
    class="bg-black"
    style="color: rgb(24, 26, 32);"
  >
    <button
      class="
      bg-forumm-blue-2 text-white relative rounded-full  px-6 py-4 text-base h-8 flex items-center hover:animate-heartbeat justify-center "
      style="background-color: rgb(55, 99, 233);"
      type="button"
    >
      <span
        class="relative z-10"
        style="color: white;"
      />
    </button>
  </div>
</div>
`;

exports[`Button Test Suite Button test 2`] = `
<div>
  <div
    class="bg-black"
    style="color: rgb(24, 26, 32);"
  >
    <button
      class="
      bg-forumm-blue-2 text-white relative rounded-full  px-6 py-4 text-base h-8 flex items-center hover:animate-heartbeat justify-center undefined"
      style="background-color: rgb(241, 241, 241);"
      type="button"
    >
      <span
        class="relative z-10"
        style="color: rgb(31, 31, 31);"
      />
    </button>
  </div>
</div>
`;
